osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -0,0 +1,168 @@
+        }
+    }
+
+    public List<TypedSecret> retrieveSecrets(final String deviceIdentification, final List<SecretType> secretTypes)
+            throws Exception {
+        try {
+            //@formatter:off
+            return secretTypes.stream()
+                    .map(secretType -> this.retrieveSecret(deviceIdentification,secretType))
+                    .collect(Collectors.toList());
+            //@formatter:on
+        } catch (final Exception exc) {
+            throw new IllegalStateException(
+                    String.format("Something went wrong retrieving secrets for device %s", deviceIdentification), exc);
+        }
+    }
+
+    @java.lang.SuppressWarnings("squid:S3655")
+    public TypedSecret retrieveSecret(final String deviceIdentification, final SecretType secretType) {
+        final Date now = new Date();
+        final Long secretId = this.secretRepository.findIdOfValidMostRecent(deviceIdentification, secretType.name(),
+                now);
+        if (secretId == null) {
+            throw new NoSuchElementException("No secret found with a valid key");
+        }
+        return this.getTypedSecret(this.secretRepository.findById(secretId).get());
+    }
+
+    private TypedSecret getTypedSecret(final DbEncryptedSecret dbEncryptedSecret) {
+        if (dbEncryptedSecret != null) {
+            final DbEncryptionKeyReference keyReference = dbEncryptedSecret.getEncryptionKeyReference();
+            if (keyReference == null) {
+                throw new IllegalStateException("Could not create encrypted secret: secret has no key reference");
+            }
+            final byte[] secretBytes = HexUtils.fromHexString(dbEncryptedSecret.getEncodedSecret());
+            final EncryptedSecret encryptedSecret = new EncryptedSecret(keyReference.getEncryptionProviderType(),
+                    secretBytes);
+            return this.createTypedSecret(dbEncryptedSecret, keyReference, encryptedSecret);
+        } else {    //Should never happen because of stream mapping in retrieveSecrets()
+            throw new IllegalStateException("Could not create typed secret for NULL secret");
+        }
+    }
+
+    private TypedSecret createTypedSecret(final DbEncryptedSecret dbEncryptedSecret,
+            final DbEncryptionKeyReference keyReference, final EncryptedSecret encryptedSecret) {
+        try {
+            final Secret decryptedSecret = this.encryptionDelegate.decrypt(encryptedSecret,
+                    keyReference.getReference());
+            final TypedSecret typedSecret = new TypedSecret();
+            typedSecret.setSecret(HexUtils.toHexString(decryptedSecret.getSecret()));
+            typedSecret.setSecretType(dbEncryptedSecret.getSecretType());
+            return typedSecret;
+        } catch (final Exception exc) {
+            throw new IllegalStateException("Could not decrypt secret (id: " + dbEncryptedSecret.getId() + ")", exc);
```

Comment:
I see many IllegalStateExceptions, although it is easy/handy to throw just a RuntimeException I would suggest to make a proper Checked Exception to make it more explicit where it can go wrong and improved exception handling.